### PR TITLE
Automatic reload of remote_servers configuration.

### PR DESCRIPTION
**Pull request by ludv1x**
Also automatic update of users' config from users.d dir introduced.

### CODE REVIEW

File: dbms/src/Storages/StorageDistributed.cpp [hunk 1]
```diff
@@ -247,7 +251,10 @@ void StorageDistributed::reshardPartitions(ASTPtr query, const String & database
 			" queries for distributed tables",
 			ErrorCodes::RESHARDING_INVALID_PARAMETERS};
 
-	std::string coordinator_id = resharding_worker.createCoordinator(cluster);
+	auto cluster = getCluster();
+
+	/// resharding_worker не нуждается во владении cluster, здесь используется только мета-информация
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageDistributed.cpp]: Please write in english.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/include/DB/Interpreters/Context.h [hunk 2]
```diff
@@ -276,8 +276,9 @@ class Context
 	  */
 	void resetCaches() const;
 
-	const Cluster & getCluster(const std::string & cluster_name) const;
-	std::shared_ptr<Clusters> getClusters() const;
+	Clusters & getClusters() const;
+	std::shared_ptr<Cluster> getCluster(const std::string & cluster_name, bool throw_on_error = true) const;
+	void setClustersConfig(ConfigurationPtr config);
```
Comment (ludv1x) [inline dbms/include/DB/Interpreters/Context.h]: We not just apply passed config, we also store config to apply it in the future (on first getClusters() call). So we need to store it (at least for clusters).
Currently, to do this we just copy shared_ptr (Poco::AutoPtr).
If we pass reference, we will must make copy of Poco::Util::AbstractConfiguration created inside ConfigReloader.

Comment (ludv1x) [inline dbms/include/DB/Interpreters/Context.h]: Reference to shared_ptr - OK.

File: dbms/src/Server/ConfigReloader.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+		FileWithTimestamp(const std::string & path_,  time_t modification_time_)
+			: path(path_), modification_time(modification_time_) {}
+
+		bool operator < (const FileWithTimestamp & rhs) const
+		{
+			return path < rhs.path;
+		}
+
+		bool operator == (const FileWithTimestamp & rhs) const
```
Comment (ludv1x) [inline dbms/src/Server/ConfigReloader.h]: Ok.
BTW. But such inconsistencies appears quite often outside Matlab-like programming world.

File: dbms/src/Common/ConfigProcessor.cpp [hunk 4]
```diff
@@ -311,22 +304,43 @@ XMLDocumentPtr ConfigProcessor::processConfig(const std::string & path_str)
...
+	///  hash function internally, and its size was prime by default.
+	Poco::AutoPtr<Poco::XML::NamePool> name_pool(new Poco::XML::NamePool(65521));
+	Poco::XML::DOMParser dom_parser(name_pool);
+
+	DocumentPtr config = dom_parser.parse(path_str);
+
+	std::vector<std::string> contributing_files;
+	contributing_files.push_back(path_str);
+
+	for (auto & merge_file : getConfigMergeFiles(path_str))
```
Comment (ludv1x) [inline dbms/src/Common/ConfigProcessor.cpp]: If I am not mistaken derived type of `auto` will be `const std::string`, so the resulting type `const std::string &` is the same as in `const auto &` case.

Comment (ludv1x) [inline dbms/src/Common/ConfigProcessor.cpp]: Ok, like small enhancement. Behavior isn't changed anyway.

**Comment by alexey-milovidov** [comment]
What if file is written non-atomically and in moment of reload, server observe a file with incorrect XML?
When will be next reload after exception?

BTW, we already have this problem with dictionaries: if file is modified not with atomic rename, there are rare race conditions, when dictionaries are not updated.

**Comment by ludv1x** [comment]
To update config on next reload modification time observed at previous iteration should differs from currently observed modification time.
So, config will be reloaded if OS sets file's modification time after all data modifications.
Otherwise, described case can occur.

To avoid this, we need to save modification time if and only if XML was loaded successfully.
But in that case if a user saves incorrect config and forget to check that changes was successfully applied, server will try to reload incorrect XML and write error logs during all remain lifetime.

So, I faced with that use case and decided to save modification time even if XML parsing failed.

**Comment by alexey-milovidov** [comment]
> So, config will be reloaded if OS sets file's modification time after all data modifications.

I am not sure about it.

> To avoid this, we need to save modification time if and only if XML was loaded successfully.

Also it is possible to reset modification time to zero on exception (means - retry next time).

> But in that case if a user saves incorrect config and forget to check that changes was successfully applied, server will try to reload incorrect XML and write error logs during all remain lifetime.

I think, it is Ok.

> So, I faced with that use case and decided to save modification time even if XML parsing failed.
It seems that, shared file lock doesn't have drawbacks. -)

We could postpone this issue until we face it.

**Comment by DmitryBe** [comment]
Do you consider an option of centralised storing remote server config in ZK?

**Comment by alexey-milovidov** [comment]
We was thinking about possibility to load any configuration sections from ZK.

**Comment by DmitryBe** [comment]
What have you decided in results? Do you plan to support ZK based config in the nearest future? If yes, approximately when?

**Comment by alexey-milovidov** [comment]
I have added it to the list, we will decide time estimate for this task at next week.

**Comment by DmitryBe** [comment]
Sounds great! Thanks a lots for quick reply.

**Pull request merged**